### PR TITLE
Dataset Service Refactoring

### PR DESCRIPTION
**Pull request by pradithya**
**What this PR does / why we need it**:

This PR was initially to implement deployment management API for log writer but it ended up revamping various components within the timber repository. 🙏 
Here are the changes within this PR

### Implement Log Writer Deployment API

This PR implements `CreateLogWriter` and `UpdateLogWriter` APIs. No DB is added yet, so the implementation is still barebone. The `LogWriter` object itself is described using following protobuf.  It currently can describe the "source" of the log but it's still missing the part describing where the logs are stored at. This will be subsequently added when implementing `LogMetadata` API.
```
// LogWriter describes details of a Log Writer
message LogWriter {
  // Project id that owns the log writer
  int64 project_id = 1;
  // Log writer's ID
  int64 id = 2;
  // Name of the log writer
  string name = 3;
  // Log writer source
  LogWriterSource source = 4;
  // TODO: Add details of where the log is stored at
  // Status of the observation service
  Status status = 10;
  // Creation timestamp
  google.protobuf.Timestamp create_time = 11;
  // Last update timestamp
  google.protobuf.Timestamp update_time = 12;
}
```

Example of API call:

To create a new prediction log writer
```bash
curl --location --request POST 'localhost:8080/v1/projects/1/log_writers' \
--header 'Content-Type: application/json' \
--data-raw '{
    "log_writer" : {
        "project_id" : 1,
        "name" : "my-prediction-log",
        "source" : {
            "type": "LOG_WRITER_SOURCE_TYPE_PREDICTION_LOG",
            "prediction_log_source" : {
                "model_id": 1,
                "model_name" : "my-model",
                "kafka" : {
                    "brokers": "kafka.mlp.svc.cluster.local",
                    "topic" : "my-prediction-log"
                }
            }
        }
    }
}'
```

to create a new router log writer
```bash
curl --location --request POST 'localhost:8080/v1/projects/1/log_writers' \
--header 'Content-Type: application/json' \
--data-raw '{
    "log_writer" : {
        "project_id" : 1,
        "name" : "my-router-log",
        "source" : {
            "type": "LOG_WRITER_SOURCE_TYPE_ROUTER_LOG",
            "router_log_source" : {
                "router_id": 1,
                "router_name" : "my-router",
                "kafka" : {
                    "brokers": "kafka.mlp.svc.cluster.local",
                    "topic" : "my-router-log"
                }
            }
        }
    }
}'
```
Since the API is still under early iteration, the E2E test is still pretty relaxed.

###  Update Observation Service Deployment API

The main change is to reduce the API complexity by hiding most of the previously exposed configuration. With this change, client can only configure 3 things: source type (kafka vs eager), kafka broker, and kafka topic. Similar to `LogWriter` APIs, it currently doesn't describe where the logs are stored at which will also be added when `LogMetadata` APIs is added. 

An example for deploying observation service which use kafka as source
```bash
curl --location --request POST 'localhost:8080/v1/projects/1/observation_services' \
--header 'Content-Type: application/json' \
--data-raw '{
    "observation_service" : {
        "project_id" : 1,
        "name" : "my-observation",
        "source" : {
            "type": "OBSERVATION_SERVICE_SOURCE_TYPE_KAFKA",
            "kafka" : {
                "brokers": "kafka.mlp.svc.cluster.local",
                "topic" : "sample"
            }
        }
    }
}'

```

### Modify Deployment Configuration Management

Observation service and log writer deployment configuration consist of 2 parts: the helm chart and default values.
The default values are intended to eliminate hardcoded configuration within dataset-service. When deploying a new log writer/observation service, dataset service will use the default values as the base configuration which will then overridden by the configuration derived from the API request. The structure of default values follow the same values.yaml of the corresponding chart.

```
// LogWriterConfig configuration for deploying log writer
type LogWriterConfig struct {
	// link to Log Writer Helm chart for deployment
	HelmChartPath string
	// Default helm values to be used when deploying log writer
	DefaultValues *values.FluentdHelmValues
}
```

```
// ObservationServiceConfig configuration for deploying observation service
type ObservationServiceConfig struct {
	// link to Observation Service Helm chart for deployment
	HelmChartPath string
	// Default helm values to be used when deploying observation service
	DefaultValues *values.ObservationServiceHelmValues
}
``` 

### Introduce `helm` package

The `helm` package contains implementation of `helm.Client` which expose helm functionality such ash `Install` and `Upgrade`. Additionally, `helm` package also contains `helm/values` package which stores the definition of helm values that were previously under `model/observation_service_helm`.

### Create Development Environment for Dataset Service

You can now easily setup a working development environment by executing following command either in the root folder or under `dataset-service` folder. 

```bash
make dev-env
```

It will create a k3d cluster and install all dependencies for testing dataset-service. The development environment is reused for e2e test so that we can ensure it's working as expected.

### Other changes

- Refactor Makefile
- Refactor README
- Remove infra/charts 

### 

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -5,45 +5,61 @@ DATASET_SVC_PATH=dataset-service
...
-# Build recipes
+# General
 # ==================================
 
-.PHONY: build-dataset-service
-build-dataset-service: version
-	@echo "Building binary..."
-	@cd ${DATASET_SVC_PATH}/cmd/${DATASET_SVC_PATH} && go build -o ./bin/${DATASET_SVC_PATH}
+.PHONY: format
+## Format all sousrce code
```
Comment (leonlnj) [inline Makefile:42]: typo 😄

File: .github/workflows/timber.yml [hunk 2]
```diff
@@ -167,21 +145,23 @@ jobs:
...
         id: setup-python
         uses: actions/setup-python@v4
         with:
           python-version: "3.9"
           architecture: x64
-      - name: Install dependencies
-        run: make install-python-ci-dependencies
-      - name: Build Dataset Service binary
-        run: make build-dataset-service
+      - uses: bufbuild/buf-setup-action@v1
```
Comment (leonlnj) [inline .github/workflows/timber.yml:154]: why is this needed?

File: dataset-service/api/dataset_service.pb.go [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
```
Comment (leonlnj) [inline dataset-service/api/dataset_service.pb.go:1]: are we going to put all proto in `/api` even though they contain the the message struct

File: dataset-service/helm/client.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf("release manifest: %v", r.Manifest)
+	return r, nil
+}
+
+// GetRelease get release name in the given namespace
+func (h helmClient) GetRelease(releaseName string, namespaceName string, actionConfig *action.Configuration) (*release.Release, error) {
```
Comment (leonlnj) [inline dataset-service/helm/client.go:127]: is there any diff between `namespaceName` and `namespace`? if not i think we can stick to `namespace`

Also this is not called anywhere for now and will eventually by the metadata controller?

File: dataset-service/helm/values/common.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	// Type of the service
+	Type string `json:"type,omitempty"`
+	// Port exposed by the service
+	ExternalPort string `json:"externalPort,omitempty"`
+	// Pod's port being mapped to the external port
+	InternalPort string `json:"internalPort,omitempty"`
+}
+
+// MerveEnvs merge 2 slices of Env and give priority for right slice
+func MerveEnvs(left []Env, right []Env) []Env {
```
Comment (leonlnj) [inline dataset-service/helm/values/common.go:94]: is this suppose to be MergeEnvs?

File: dataset-service/server/server.go [hunk 6]
```diff
@@ -76,7 +77,15 @@ func NewServer(configFiles []string) (*Server, error) {
 	}
 
 	// Creating mux for gRPC gateway. This will multiplex or route request to different gRPC service.
-	mux := runtime.NewServeMux()
+	mux := runtime.NewServeMux(runtime.WithMarshalerOption(runtime.MIMEWildcard, &runtime.JSONPb{
```
Comment (leonlnj) [inline dataset-service/server/server.go:80]: is this to prevent the server to init with zero value?

File: infra/local/dataset-service/README.md [hunk 7]
```diff
@@ -0,0 +1,11 @@
+## Setup Local Development Environment
+
+### Setup Cluster
+1. create k3d cluster
+2. deploy kafka into cluster
+3. deploy mlp into cluster and expose
+4. deploy postgresql into cluster and expose
```
Comment (leonlnj) [inline infra/local/dataset-service/README.md:7]: 4 and 5 are for future addition right?

File: tests/Makefile [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+## Install python dependencies
+.PHONY: dep
+dep:
+	pip install -r requirements.txt
+
+## Setup e2e test environment and dependencies
+.PHONY: setup-e2e
+setup-e2e:
+	$(MAKE) -C ../ dep
```
Comment (leonlnj) [inline tests/Makefile:30]: should this be `setup-e2e` instead of `dep`?

File: dataset-service/controller/logwriter_controller.go [hunk 9]
```diff
@@ -53,40 +53,54 @@ func (l LogWriterController) GetLogWriter(
...
+	project, err := l.appCtx.Services.MLPService.GetProject(projectID)
 	if err != nil {
+		log.Errorf("error finding project: %v", err)
 		return nil, err
 	}
 
-	// TODO: Implement method
-	log.Info("Called caraml.upi.v1.DatasetService/CreateLogWriter")
-	response := &timberv1.CreateLogWriterResponse{}
+	logWriter, err := l.appCtx.Services.LogWriterService.Create(project.Name, r.LogWriter)
```
Comment (leonlnj) [inline dataset-service/controller/logwriter_controller.go:68]: i cant remember but does this block till helm installed successfully? 

not a problem for now, just to clarify if the client might wait a long time (if there is auto scaling of node event) before getting a response

File: dataset-service/config/config.go [hunk 10]
```diff
@@ -11,35 +12,69 @@ import (
...
-	// GCPProject specifies the GCP project where BQ logs will be written to via Fluentd
-	GCPProject string
-	// ObservationServiceImageTag specifies tag of image to be used when deploying Observation Service via Dataset Service APIs
-	ObservationServiceImageTag string
-	// FluentdImageTag specifies tag of image to be used when deploying Fluentd via Dataset Service APIs
-	FluentdImageTag string
+// CommonDeploymentConfig configuration common to both observation-service and log writer deployment
+type CommonDeploymentConfig struct {
+	// environment type of the deployment
+	EnvironmentType string `default:"local"`
```
Comment (leonlnj) [inline dataset-service/config/config.go:42]: this is only used sentry config label for now. Will there be more extensive use of this? I think Turing is checking the env matches the environment api. Im not too sure of the usefulness of this, but basically this input can be anything or empty for now and the app works

File: dataset-service/proto/caraml/timber/v1/log_metadata.proto [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    string id = 1;
+    // Name of the producer, dependent on the type of the log.
+    string name = 2;
+    // Name of the CaraML project that hosts the producer.
+    string project = 3;
+}
+
+// LogMetadata is an entity/metadata in Dataset Service that represents an append-only
+// data produced by ingesting the observation, prediction, or router logs
+message LogMetadata {
```
Comment (leonlnj) [inline dataset-service/proto/caraml/timber/v1/log_metadata.proto:32]: who is responsible for generating this entry? 

I assume its will be the respective obs/logwriter controller upon CRUD of a service? since there is no create api

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto [hunk 12]
```diff
@@ -194,35 +194,35 @@ message GetObservationServiceRequest {
...
+    ObservationService observation_service = 1;
 }
 
 // Request message for UpdateObservationService method.
 message UpdateObservationServiceRequest {
-    // The CaraML project ID to retrieve log resource from.
+    // The CaraML project ID to write log resource from.
     int64 project_id = 1;
-    // The ID of Observation Service to update.
+    // The ID of Log Writer to update.
```
Comment (leonlnj) [inline dataset-service/proto/caraml/timber/v1/dataset_service.proto:218]: actually is the ID being used anywhere? The update should just update the helm chart according (which will redeploy fluentd) and I didnt see anywhere the ID is used.

Also why is this log writer 😅

File: dataset-service/proto/caraml/timber/v1/log_writer.proto [hunk 13]
```diff
@@ -3,17 +3,65 @@ syntax = "proto3";
...
+  KafkaConfig kafka = 3;
+}
+
+message RouterLogSource {
+  // id of the router producing the router logs
+  int64 router_id = 1;
+  // name of the router producing the router logs
+  string router_name = 2;
+  // kafka source configuration where the router logs are located
+  KafkaConfig kafka = 3;
```
Comment (leonlnj) [inline dataset-service/proto/caraml/timber/v1/log_writer.proto:47]: will users be calling the dataset service api directly? 

The kafka topic is in a predefined format in Merlin/Turing and unknown to the user and the broker is configured app wide. We can fix the pattern in UI (and maybe SDK if we have one)

Not an issue to make it generic to support a generic way of using Timber, just highlighting user might not know this upfront.

File: dataset-service/service/observation_service.go [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+func setLogConsumerConfig(
+	val *values.ObservationServiceHelmValues,
+	svc *timberv1.ObservationService,
+) (*values.ObservationServiceHelmValues, error) {
+	switch svc.GetSource().GetType() {
+	case timberv1.ObservationServiceSourceType_OBSERVATION_SERVICE_SOURCE_TYPE_EAGER:
+		return nil, fmt.Errorf("source type (eager) is currently unsupported")
+	case timberv1.ObservationServiceSourceType_OBSERVATION_SERVICE_SOURCE_TYPE_KAFKA:
+		val.ObservationService.APIConfig.LogConsumerConfig.Kind = osconfig.LoggerKafkaConsumer
+		val.ObservationService.APIConfig.LogConsumerConfig.KafkaConfig.Topic = svc.Source.Kafka.Topic
```
Comment (leonlnj) [inline dataset-service/service/observation_service.go:130]: as a todo, maybe we should add a validator or app configuration of allowed pattern, similiar to `dataset-service/bq/name.go`, so that we dont create service that read in anywhere else unexpected

**Comment by leonlnj** [comment]
that was long 😅 

LVGTM, huge improvement 🙏 , as a whole no major comment just some minor questions

File: .github/workflows/timber.yml (hunk 2 above)
Comment (pradithya) [reply .github/workflows/timber.yml:154]: Initially, I added `make generate` but reverted it, thus this is not needed anymore.

File: dataset-service/api/dataset_service.pb.go (hunk 3 above)
Comment (pradithya) [reply dataset-service/api/dataset_service.pb.go:1]: Yes, that's intentional, so that the import path is not too deep.

File: dataset-service/helm/client.go (hunk 4 above)
Comment (pradithya) [reply dataset-service/helm/client.go:127]: > is there any diff between namespaceName and namespace? if not i think we can stick to namespace

Agree, will fix it.

> Also this is not called anywhere for now and will eventually by the metadata controller?

Yeah, it's not being used. It could be used by the read/list APIs.

File: dataset-service/README.md [hunk 15]
```diff
@@ -1,32 +1,117 @@
...
 
-To generate all code and documentation
+```bash
+make setup
 ```
-make
+
+To build dataset service binary execute
+
+``` bash
```
Comment (terryyylim) [inline dataset-service/README.md:26]: nit: Extra space between ``` and bash

File: dataset-service/server/server.go (hunk 6 above)
Comment (pradithya) [reply dataset-service/server/server.go:80]:
```
		MarshalOptions: protojson.MarshalOptions{
			UseProtoNames: true,
		},
```
is to make sure the json response field naming follow protobuf name. So instead of `observationService` it should be `observation_service`.

```
		UnmarshalOptions: protojson.UnmarshalOptions{
			DiscardUnknown: true,
		},
```

is to ignore fields in request that are not part of protobuf definition.

File: infra/local/dataset-service/README.md (hunk 7 above)
Comment (pradithya) [reply infra/local/dataset-service/README.md:7]: Yeah, 4 and 5 are not implemented yet

File: dataset-service/config/config.go (hunk 10 above)
Comment (pradithya) [reply dataset-service/config/config.go:42]: I think it's better to delete it for now. I don't think we'll use it extensively.

File: dataset-service/config/local.yaml [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+      app: dataset-service
+commonDeploymentConfig:
+  environmentType: local
+  kubeConfig: /tmp/kubeconfig-timber-dev.yaml
+  bqConfig:
+    gcpProject: my-gcp-project
+    bqDatasetPrefix: caraml
+    observationBqTablePrefix: os
+observationServiceConfig:
+  # You can replace   helmChartPath to point to local directory
```
Comment (terryyylim) [inline dataset-service/config/local.yaml:30]: nit: extra space before `helmChartPath`

File: dataset-service/proto/caraml/timber/v1/log_metadata.proto (hunk 11 above)
Comment (pradithya) [reply dataset-service/proto/caraml/timber/v1/log_metadata.proto:32]: > I assume its will be the respective obs/logwriter controller upon CRUD of a service? since there is no create API

Yes, you are right! it will be created when a new observation service or log writer is created.

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto (hunk 12 above)
Comment (pradithya) [reply dataset-service/proto/caraml/timber/v1/dataset_service.proto:218]: > actually is the ID being used anywhere? The update should just update the helm chart according (which will redeploy fluentd) and I didnt see anywhere the ID is used.

The grpc gateway complained if the ID is removed. I think internally we should check whether the `ID` is consistent with the payload.

File: dataset-service/service/observation_service.go (hunk 14 above)
Comment (pradithya) [reply dataset-service/service/observation_service.go:130]: Yeah, we'll need to add request validation and config validation into our backlog.

File: dataset-service/helm/client.go [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	settings := cli.New()
+	chartPathOption := action.ChartPathOptions{}
+	chartPath, err := chartPathOption.LocateChart(chartPath, settings)
+	if err != nil {
+		return nil, fmt.Errorf("error locating chart %s, %w", chartPath, err)
+	}
+
+	c, err = loader.Load(chartPath)
```
Comment (terryyylim) [inline dataset-service/helm/client.go:63]: `Load` handles both a remote file path (i.e helm-charts repo) or local directory right?

File: dataset-service/helm/client.go [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error initializeConfig: %w", err)
+	}
+	get := action.NewStatus(actionConfig)
+
+	return get.Run(releaseName)
+}
+
+// initializeConfig initialize action config
+func (h *helmClient) initializeConfig(actionConfig *action.Configuration, namespaceName string) (*action.Configuration, error) {
```
Comment (terryyylim) [inline dataset-service/helm/client.go]: nitpicking my own implementation of this function previously lol - I guess `namespace` rather than `namespaceName` would suffice, similar to Leon's comment above.

File: dataset-service/helm/client_test.go [hunk 19]
```diff
@@ -0,0 +1,193 @@
+package helm
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/caraml-dev/timber/common/log"
```
Comment (terryyylim) [inline dataset-service/helm/client_test.go:7]: nit: timber import in separate section?

File: dataset-service/helm/values/common.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+// AutoscalingConfig Autoscaling configuration
+type AutoscalingConfig struct {
+	// Enable/disable autoscaling flag
+	Enabled bool `json:"enabled,omitempty"`
+	// Minimum number of replicas for the deployment
+	MinReplicas int `json:"minReplicas,omitempty"`
+	// Maximum number of replicas for the deployment
+	MaxReplicas int `json:"maxReplicas,omitempty"`
+	// CPU target utilization in percentage
+	TargetCPUUtilizationPercentage int `json:"targetCPUUtilizationPercentage,omitempty"`
```
Comment (terryyylim) [inline dataset-service/helm/values/common.go:46]: Shall we expose `TargetMemoryUtilizationPercentage` too?

File: dataset-service/helm/values/fluentd.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+	// Annotations to be added to the deployment
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// Additional labels to be added to the deployment
+	ExtraLabels map[string]string `json:"extraLabels,omitempty"`
+	// Number of replica
+	ReplicaCount int `json:"replicaCount,omitempty"`
+	// Resource configuration (i.e. CPU & Memory requests and limits)
+	Resources ResourcesConfig `json:"resources,omitempty"`
+	// kubernetes service account to be mounted to the pod created by the deployment
+	KubernatesServiceAccount string `json:"kubernatesServiceAccount,omitempty"`
```
Comment (terryyylim) [inline dataset-service/helm/values/fluentd.go:18]: typo: 
```
KubernetesServiceAccount string `json:"kubernetesServiceAccount,omitempty"`
```

File: dataset-service/helm/values/fluentd.go [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+	// Google Service account to be mounted to the pod created by the deployment
```
Comment (terryyylim) [inline dataset-service/helm/values/fluentd.go:19]: nit: to be mounted on the pod

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto [hunk 23]
```diff
@@ -194,35 +194,35 @@ message GetObservationServiceRequest {
...
 message GetObservationServiceResponse {
     // Details of single Observation Service deployed by Dataset Service.
-    ObservationServiceConfig observation_service = 1;
+    ObservationService observation_service = 1;
 }
 
 // Request message for CreateObservationService method.
 message CreateObservationServiceRequest {
-    // The CaraML project ID to retrieve log resource from.
+    // The CaraML project ID to write log resource from.
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/dataset_service.proto:202]: nit: write log resource to

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto [hunk 24]
```diff
@@ -194,35 +194,35 @@ message GetObservationServiceRequest {
...
     int64 project_id = 1;
     // Details of single Observation Service to be deployed by Dataset Service.
-    ObservationServiceConfig observation_service = 2;
+    ObservationService observation_service = 3;
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/dataset_service.proto:205]: Do we need to bump this value? I think we can keep it as `2` since we're not officially released yet?

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto [hunk 25]
```diff
@@ -194,35 +194,35 @@ message GetObservationServiceRequest {
...
 message CreateObservationServiceResponse {
     // Details of single Observation Service deployed by Dataset Service.
-    ObservationServiceResponse observation_service = 1;
+    ObservationService observation_service = 1;
 }
 
 // Request message for UpdateObservationService method.
 message UpdateObservationServiceRequest {
-    // The CaraML project ID to retrieve log resource from.
+    // The CaraML project ID to write log resource from.
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/dataset_service.proto:216]: nit: write log resource to

File: dataset-service/proto/caraml/timber/v1/log_writer.proto [hunk 26]
```diff
@@ -3,17 +3,65 @@ syntax = "proto3";
...
+  int64 id = 2;
+  // Name of the log writer
+  string name = 3;
+  // Log writer source
+  LogWriterSource source = 4;
+  // TODO: Add details of where the log is stored at
+  // Status of the observation service
+  Status status = 10;
+  // Creation timestamp
+  google.protobuf.Timestamp create_time = 11;
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/log_writer.proto:64]: Would `created_at` be clearer?

File: dataset-service/proto/caraml/timber/v1/log_writer.proto [hunk 27]
```diff
@@ -3,17 +3,65 @@ syntax = "proto3";
...
+  // Last update timestamp
+  google.protobuf.Timestamp update_time = 12;
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/log_writer.proto:66]: Would `updated_at` be clearer?

File: dataset-service/proto/caraml/timber/v1/observation_service.proto [hunk 28]
```diff
@@ -3,56 +3,44 @@ syntax = "proto3";
...
-    // Configuration for Observation Service data sink.
-    ObservationServiceDataSink sink = 3;
+    ObservationServiceSource source = 4;
+    // TODO: Add details of where the log is stored at
+    // Status of the observation service
+    Status status = 10;
+    // Creation timestamp
+    google.protobuf.Timestamp create_time = 11;
+    // Last update timestamp
+    google.protobuf.Timestamp update_time = 12;
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/observation_service.proto:44]: Would `updated_at` be clearer?

File: dataset-service/proto/caraml/timber/v1/observation_service.proto [hunk 29]
```diff
@@ -3,56 +3,44 @@ syntax = "proto3";
...
     // Configuration for pull-based Observation Service data source.
-    ObservationServiceDataSource source = 2;
-    // Configuration for Observation Service data sink.
-    ObservationServiceDataSink sink = 3;
+    ObservationServiceSource source = 4;
+    // TODO: Add details of where the log is stored at
+    // Status of the observation service
+    Status status = 10;
+    // Creation timestamp
+    google.protobuf.Timestamp create_time = 11;
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/observation_service.proto:42]: Would `created_at` be clearer?

File: dataset-service/proto/caraml/timber/v1/log_writer.proto [hunk 30]
```diff
@@ -3,17 +3,65 @@ syntax = "proto3";
...
+  // Project id that owns the log writer
+  int64 project_id = 1;
+  // Log writer's ID
+  int64 id = 2;
+  // Name of the log writer
+  string name = 3;
+  // Log writer source
+  LogWriterSource source = 4;
+  // TODO: Add details of where the log is stored at
+  // Status of the observation service
```
Comment (terryyylim) [inline dataset-service/proto/caraml/timber/v1/log_writer.proto:61]: typo: Status of the log writer

File: dataset-service/proto/caraml/timber/v1/log_writer.proto (hunk 13 above)
Comment (pradithya) [reply dataset-service/proto/caraml/timber/v1/log_writer.proto:47]: > will users be calling the dataset service API directly? 

To create the log writer? I am envisioning that eventually no user interaction is required. But I think for MVP, we can expose this in UI.

File: dataset-service/service/log_writer_service.go [hunk 31]
```diff
@@ -0,0 +1,201 @@
+package service
+
+import (
+	"errors"
+	"fmt"
+
+	timberv1 "github.com/caraml-dev/timber/dataset-service/api"
```
Comment (terryyylim) [inline dataset-service/service/log_writer_service.go:7]: nit: timber imports in separate section? Similar comment everywhere else

File: tests/README.md [hunk 32]
```diff
@@ -1,48 +1,30 @@
...
-#### <b>Run E2E tests</b>
----
 
-- Builds Dataset Service binary
-- Starts k3d cluster via Docker Compose
-- Run pytests against k3d cluster
+Lastly, execute the end to end test
 
 ```bash
 make e2e
```
Comment (terryyylim) [inline tests/README.md:30]: Should this be `make test`?

File: tests/e2e/test_dataset_service.py [hunk 33]
```diff
@@ -1,99 +1,210 @@
-from datetime import datetime
```
Comment (terryyylim) [inline tests/e2e/test_dataset_service.py:1]: 😄 

image

**Review by terryyylim — COMMENTED** [review]
Thank you for the major re-organisation @pradithya ! Left couple of comments 🙏🏻

File: dataset-service/config/config.go [hunk 34]
```diff
@@ -11,35 +12,69 @@ import (
...
+	// BQ Config
+	BQConfig *BQConfig
+}
+
+// BQConfig BigQuery configuration
+type BQConfig struct {
+	// GCPProject specifies the GCP project where BQ logs will be written to via FluentdHelmValues
+	GCPProject string
+	// BigQuery dataset prefix
+	BQDatasetPrefix string `default:"caraml"`
```
[the remark this answers is not in the document]
Comment (pradithya) [reply dataset-service/config/config.go:52]: I think the config would pretty easy to change if we add different storage support. For now, I am keeping it simple.

File: README.md [hunk 35]
```diff
@@ -1,54 +1,23 @@
...
+### Dataset Service
 
-#### a. Build dependencies
-```bash
-# Build Fluentd image
-make build-fluentd-image
-```
-
-#### b. Setup dependencies
+Dataset service is the orchestrator within Timber ecosystem. It manages deployment of dataset service and observation service.
```
[the remark this answers is not in the document]
Comment (pradithya) [reply README.md:11]: It's what happens when I lack sleep.

File: dataset-service/helm/client.go (hunk 17 above)
Comment (pradithya) [reply dataset-service/helm/client.go:63]: Yes, it can handle local path (`file:///path/to/chart`) or remote URL

File: dataset-service/helm/client.go (hunk 18 above)
Comment (pradithya) [reply dataset-service/helm/client.go]: Yep, renamed it to `namespace` for consistency

File: dataset-service/helm/values/common.go (hunk 20 above)
Comment (pradithya) [reply dataset-service/helm/values/common.go:46]: Yes, added

File: dataset-service/proto/caraml/timber/v1/dataset_service.proto (hunk 24 above)
Comment (pradithya) [reply dataset-service/proto/caraml/timber/v1/dataset_service.proto]: Reverted to 2

**Review by terryyylim — APPROVED** [review]

**Comment by pradithya** [comment]
@leonlnj I've addressed your feedback. Let me know if there's anything left

**Review by leonlnj — APPROVED** [review]
thanks, LGTM!
